### DISCUSSION TITLE
Turn query result to application/sparqlresult+json

### DISCUSSION THREAD

**Discussion by BenjaminHofstetter** [oxigraph/oxigraph#627] [Q&A]
I'm using the JavaScript binding for Oxigraph. How can I convert the result of a SELECT query to application/sparqlresult+json format? I did not find a way todo it. I tried doing it myself, but I'm having trouble maintaining the correct order of head.vars in all cases (with UNIONs).

**Comment by Tpt** [comment]
Hi!

That's a great question! I don't think it's currently possible with the current JS api that returns an iterator of maps and it's a shame.
I see three way forwards:
1. With the current Oxigraph JS API: `Map` should keep insertion order so if you iterate on all the results and merge the variable list of each result you should get the proper ordering of result variable but you will still miss the variables that are always undefined. But's very cumbersome to write. I guess it's more or less what you are already doing?
2. With a backward compatible change: I can introduce a `query_with_results` method that would execute a SPARQL query and returns its results as a application/sparqlresult+json object. But it's sad to duplicate the `query` and it would be great to support also the other result formats.
3. With a breaking change: I can change the return type of `query` to return an object that would have a method `serialize` that allows to serialize query results and would still allow iteration yielding solutions or quads.

The first one is a quick fix, the third one is probably the way to go long-term wise. The current JS API is a quick&dirty experiment I made, I would love to get it to a better shape and help would be very welcomed in this area, I am not a JS specialist. I have opened #628 about it.

BTW, congrats for https://github.com/zazuko/vscode-sparql-notebook/issues/37#issuecomment-1713709660

**Reply by BenjaminHofstetter** [reply]
Yes I do something like mentioned in 1. To be honest, I prefer the Bindings Integrator more than application/sparqresult+json structure. 

Thank you for the sparql notebook flowers. Oxigraph is the biggest part of this feature :-) thank you for it.
